### PR TITLE
Tessellator: Improve logic when two holes share the same vertex with the polygon

### LINKED ISSUE #13841: Unable to Tessellate shape for a valid Polygon according to GDAL/OGR and PostGIS
### Description

Hello,

We recently have some issues with a Polygon geometry and the following error: "Unable to Tessellate shape” when we try to create a new document with the `geo_shape` type.

The error sends by Elasticsearch is:

```
{'error': {'root_cause': [{'type': 'document_parsing_exception',
    'reason': '[1:7350] failed to parse field [location] of type [geo_shape]'}],
  'type': 'document_parsing_exception',
  'reason': '[1:7350] failed to parse field [location] of type [geo_shape]',
  'caused_by': {'type': 'illegal_argument_exception',
   'reason': 'Unable to Tessellate shape. Possible malformed shape detected.'}},
 'status': 400}
```

Note that this is due to a Polygon which is valid according to other GIS tools such as GDAL/OGR or PostGIS.

## How to reproduce

I have this Elastic mapping

```
{
    "mappings": {
        "properties": {
            "location": {
                "type": "geo_shape"
            }
        }
    }
}
```

Here a GeoJSON file with the Polygon which raises the error.

[invalid-polygon.json](https://github.com/user-attachments/files/17209828/invalid-polygon.json)

### Version and environment details

I tried to reproduce this issue for different Elasticsearch versions: 7.17.24 and 8.15.1.

Elasticsearch 7.17.24 with Lucene 8.11.3

```
{"number": "7.17.24",
 "build_flavor": "default",
 "build_type": "tar",
 "build_hash": "fcf25fff740db6ab3ed5d145c58d70e4c3528ea7",
 "build_date": "2024-09-05T07:34:51.812485320Z",
 "build_snapshot": false,
 "lucene_version": "8.11.3"}
```

Elasticsearch 8.15.1 and Lucene 9.11.1

```
{"number": "8.15.1"
 "build_flavor": "default",
 "build_type": "tar",
 "build_hash": "253e8544a65ad44581194068936f2a5d57c2c051",
 "build_date": "2024-09-02T22:04:47.310170297Z",
 "build_snapshot": false,
 "lucene_version": "9.11.1"}
```

About my OS, I'm on MacOS Sequoia 15.0. The JVM used is the one which is included when I downloaded the `.tar.gz` from the Elastic website.

```
openjdk version "22.0.1" 2024-04-16
OpenJDK Runtime Environment (build 22.0.1+8-16)
OpenJDK 64-Bit Server VM (build 22.0.1+8-16, mixed mode, sharing)
```

Thanks!

### PR DESCRIPTION
**Pull request by iverase**
One of the situation with David Eberly's algorithm for finding  a bridge between a hole and outer polygon was failing was the case of a polygon sharing a vertex with the outer polygon. To fixed that we added a step to find shared vertex between the hole and the polygon.

That worked well except on the situations where more than one hole was sharing the same vertex with the outer polygon. In those cases, we might have more than one vertex to choose from and we need to make sure we pick the right one. The current algorithm test if the line joining the previous vertex if the outer polygon band the next vertex from the hole intersects the polygon. This worked well in many cases but it is not generic as it depends on the topology of the union.

This commit tries to improve the algorithm by:

1) If there is only one vertex, then there is not further checks and that's the one used.
2) if there is a common vertex, it first compute the signed area of the join, if they have different sign, it chooses the negative one as that's the convex union. If they have the same sign, it computes the angle of the join and chooses the smallest angle.

I added a few test to check the different combinations and the new algorithm seems resilient to them.

In addition I clean the tessellator class by removing some useless final declarations for static methods and clean up the visibility of some internal classes (no point of protected inner classes on a final class) and remove unused methods. I clean up the tests too.

fixes https://github.com/apache/lucene/issues/13841

### CODE REVIEW

**Review by craigtaverner — CHANGES_REQUESTED** [review]
Looks great, except for the method name `fetchSharedVertex` which I really feel needs to change.

I did notice that two of the new geojson files were actually WKT, but think that file names are not really important here.

File: lucene/core/src/java/org/apache/lucene/geo/Tessellator.java [hunk 1]
```diff
@@ -390,12 +377,108 @@ private static final void eliminateHole(
     }
   }
 
+  /** Choose a common vertex between the polygon and the hole if it exists, otherwise return null */
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/geo/Tessellator.java]: makes sense, renamed to `maybeMergeHoleWithSharedVertices`

File: lucene/core/src/test/org/apache/lucene/geo/TestTessellator.java [hunk 2]
```diff
@@ -928,12 +848,42 @@ public void testComplexPolygon55() throws Exception {
   public void testComplexPolygon56() throws Exception {
     String geoJson = GeoTestUtil.readShape("github-12352-2.geojson.gz");
     Polygon[] polygons = Polygon.fromGeoJSON(geoJson);
-    for (Polygon polygon : polygons) {
-      List<Tessellator.Triangle> tessellation =
-          Tessellator.tessellate(polygon, random().nextBoolean());
-      assertEquals(area(polygon), area(tessellation), 0.0);
-      // don't check edges as it takes several minutes
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/test/org/apache/lucene/geo/TestTessellator.java]: good catch, I reverted.

**Review by craigtaverner — APPROVED** [review]
LGTM

**Comment by ayinresh** [comment]
Heads up @iverase + @craigtaverner - I believe this introduced this regression https://github.com/apache/lucene/issues/15205

**Pull request merged**